### PR TITLE
Added code to synchronise start and endchat call

### PR DESCRIPTION
**Pull request by mithunjohny**
# PR Details
Issue: Race condition between startChat() and endChat() operations causing state corruption and inconsistent chat behavior.

Description
Problem to be solved: Race conditions occur when startChat() and endChat() operations execute concurrently, leading to:

State corruption where resources are cleaned up while still being initialized
Inconsistent chat session states that can leave conversations in undefined conditions
Resource leaks where conversations are not properly cleaned up
Unpredictable user experience in scenarios involving rapid successive chat operations
The core issue is that both operations modify shared state (conversation objects, chat tokens, timers) without proper synchronization, creating a window for interference between operations.

Solution Proposed
Mutex Pattern with Operation Queue Implementation:

Implemented a comprehensive operation serialization mechanism that ensures startChat() and endChat() operations execute in a mutually exclusive manner:

Core Components:

chatOperationInProgress: boolean - Tracks if an operation is currently executing
pendingOperations: Array<() => void> - Queue for pending operations
executeWithLock<T>() - Core mutex method that wraps operations
processNextOperation() - Queue processor for pending operations
Implementation Pattern:

Modified public startChat() and endChat() methods to use executeWithLock()
Created internal internalStartChat() and internalEndChat() methods for actual implementation
Operations are queued when another operation is in progress
Queue processes in FIFO order maintaining operation sequence
Error Handling:

Failed operations don't break the mutex state
Queue continues processing subsequent operations after failures
Proper cleanup ensures mutex is released in all scenarios

## Test cases and evidence

_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

### Sanity Tests

- [ ] You have tested all changes in Popout mode
- [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
- [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y

- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

***Please provide justification if any of the validations has been skipped.***

### CODE REVIEW

File: __tests__/OmnichannelChatSDKMutex.web.spec.ts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            
+            expect(result).toBe('result');
+            expect(mockOperation).toHaveBeenCalledTimes(1);
+            expect((omnichannelChatSDK as any).chatOperationInProgress).toBe(false);
+            expect((omnichannelChatSDK as any).pendingOperations).toHaveLength(0);
+        });
+
+        it('should queue operation when another operation is in progress', async () => {
+            const firstOperation = jest.fn().mockImplementation(() =>
+                new Promise(resolve => setTimeout(() => resolve('first'), 100))
```
[the remark this answers is not in the document]
Comment (mithunjohny) [reply __tests__/OmnichannelChatSDKMutex.web.spec.ts:56]: I think the fakeTimers are used to test timers in the actual code right, Here i am just creating a mockImplementation of startChat and mocking the internals with a real timer

**Review by Sonam-Ligade — APPROVED** [review]

File: src/OmnichannelChatSDK.ts [hunk 2]
```diff
@@ -227,6 +231,44 @@ class OmnichannelChatSDK {
         loggerUtils.setRequestId(this.requestId, this.ocSdkLogger, this.acsClientLogger, this.acsAdapterLogger, this.callingSdkLogger, this.amsClientLogger, this.ic3ClientLogger);
     }
 
+    /**
+     * Executes an operation with mutual exclusion to prevent race conditions
+     * between startChat and endChat operations
+     */
+    private async executeWithLock<T>(operation: () => Promise<T>): Promise<T> {
```
Comment (SamuelBishop) [inline src/OmnichannelChatSDK.ts:238]: What about a scenario where there are server problems with the start chat API via chatSDK? i.e. - startChat hangs (gets stuck). Is there a timeout scenario for this on the backend side so we do not accidentally block the pendingOperations queue?

I think for now this is fine, because only startChat and endChat are executing with the lock, and one cannot end a chat before starting it, but the reverse should also be considered. What if endChat hangs, and then the next conversation cannot call start chat because the mutex is locked on the previous endChat operation?

One way to resolve such problems might be to give executeWithLock a timeout mechanism, that way eventually the pendingOperations queue will clear, regardless of hanging.
